simulation/simple_sensor_simulator/include/simple_sensor_simulator/vehicle_simulation/vehicle_model/sim_model_delay_steer_acc_geared.hpp

```diff
@@ -62,6 +62,7 @@ class SimModelDelaySteerAccGeared : public SimModelInterface
     VX,
     STEER,
     ACCX,
+    Z,
```

Comment:
The vehicle_model in scenario_simulator_v2 is a hard copy of the one included in autowarefoundation/autoware. This is done to keep the differences between scenario_simulator_v2 and simple_planning_simulator small. Therefore, we do not want to make any changes to vehicle_model for scenario_simulator_v2.
Please fix the issue without modifying vehicle_model.

Reply:
@yamacir-kit 
Please take a look at this part of the description:

> ### [Oz position in SimModelInterface](https://github.com/tier4/scenario_simulator_v2/pull/1226/commits/46490d6dd7239416c39905652878692b19dee53b)
> - This change is necessary for FollowTrajectoryAction to work properly. Entities move along lanelets, where pitch and Oz are taken into account. If FollowTrajectoryAction is to overwrite Ego positions, the position in Oz cannot just be ignored.
> - Currently, the virtual `getZ()` method has been implemented, which returns 0 for every model except the one currently in use.
> - The position in Oz is set only when the position is overwritten, it is not calculated in any way. To be more precise, if the `update()` in SimModelInterface is used, everything still works on 2D, and the position in Oz is only used during `overwrite()` - that is, only by FollowTrajectoryAction.
> - I believe that all SimModelInterface should be adapted to 3D, due to the fact that all the rest of the code in ss2 currently uses 3D and Autoware uses 3D.

The issue that occurs is related to the lack of consideration of the `Oz` axis in the calculations. As a result, after each update the entire position (in the `Ox` and `Oy` axes also) is inconsistent with that which was overwritten. 

As far as I understand, on the `Autoware` side both `Oz` and pitch are taken into account. Recently, complete consideration of both `Oz` and pitch was introduced on the `scenario_simulator_v2` side. 

The issue I encountered is related to the fact that in the [SimModelInterface](https://github.com/tier4/scenario_simulator_v2/pull/1226/commits/46490d6dd7239416c39905652878692b19dee53b) axis `Oz` is not taken into account. This causes that during communication `EgoEntitySimulation->SimModelInteface->Autoware` there is a conversion between `3D->2D->3D` which introduces inaccuracies (position differences), and a simple alignment to `lanelet` (matching `Oz` position to `lanelet`) does not solve the issue, because the values in `Ox` and `Oy` are also affected and not the same. 

Therefore, the only alternative solution I can propose for the proper operation of `FollowTrajectotryAction` for Ego (when `isControlledBySimulator()==true`) is to completely ignore the position received from `Autoware`. This implies that the status in `EgoEntitySimulation` will not be calculated using `SimModelInterface` - it will simply remain overwritten without “getting the values from Autoware”.

However, I believe that it is inevitable to introduce the `Oz` axis in `SimModelInteface` to make the positions passed between `scenario_simulator_v2` and `Autoware` the same.

Reply:
@dmoszynski 
I understand that this change is essential due to the request for accuracy.

However, since simple_planning_simulator is still the one most often used by Autoware's planner development team, its vehicle_model and ssv2's vehicle_model should be kept equivalent.

Therefore, I will now propose to simple_planning_simulator to change about `getZ` in this pull request. After it is confirmed that there is no problem with the proposal, I will ask you to send a pull request to autowarefoundation/autoware. I will then justify the changes in this pull request by saying “import the changes in autowarefoundation/autoware”.

Thanks for your detailed explanation.

Reply:
> I will now propose to simple_planning_simulator to change about getZ in this pull request. After it is confirmed that there is no problem with the proposal, I will ask you to send a pull request to autowarefoundation/autoware. 

Thank you so much, then I'm waiting for information about the ability to create a PR.

Reply:
@dmoszynski 
I am revisiting the changes required by this pull request in order to propose changes to the planner development team. So, I am wondering about the following:

- IDX::Z is not involved in the calculations inside SimModel.
- IDX::Z exists only to transmit the value of the Z coordinate from EgoEntitySimulation::overwrite to EgoEntitySimulation::getCurrentPose.
- Therefore, IDX::Z does not need to be included in SimModel and can be replaced by a data member `z` in the EgoEntitySimulation class.

Please let me know if the above is correct. If it is correct, there is no need to change simple_planning_simulator, so please change the code as above.

Reply:
@yamacir-kit 
Thank you for pointing this out, I developed this code more than a month ago, I will take a closer look at it and let you know why I concluded ( and if right) that such changes are required.

Reply:
@yamacir-kit 
I've developed a [solution with no need to modify](https://github.com/tier4/scenario_simulator_v2/pull/1226/commits/6024371b9b0845cfc3ca7e47e6874b9de00ef5fb) `SimModelInterface`. I've performed tests using the scenarios from the PR description - it confirms the solution to the issue - videos below. I run evaluator for the current version of the branch, I will share the results on the jira ticket. 
#### Scenario with slopes - EGO

https://github.com/tier4/scenario_simulator_v2/assets/121798334/23fc7b5c-5292-41a4-a7ce-bfba5289d99a


#### [**autoware.yaml](https://github.com/tier4/scenario_simulator_v2/blob/master/test_runner/scenario_test_runner/scenario/RoutingAction.FollowTrajectoryAction-autoware.yaml) - EGO

https://github.com/tier4/scenario_simulator_v2/assets/121798334/9949dd87-e76f-4049-9ae5-0dc04c7b1a9a